bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java

```diff
@@ -578,12 +581,13 @@ void addEntry(final long ledgerId, byte[] masterKey, final long entryId, ByteBuf
                 headerBuilder.setPriority(DEFAULT_HIGH_PRIORITY_VALUE);
             }
 
-            byte[] toSendArray = toSend.toArray();
             AddRequest.Builder addBuilder = AddRequest.newBuilder()
                     .setLedgerId(ledgerId)
                     .setEntryId(entryId)
-                    .setMasterKey(ByteString.copyFrom(masterKey))
-                    .setBody(ByteString.copyFrom(toSendArray));
+                    .setMasterKey(UnsafeByteOperations.unsafeWrap(masterKey))
+                    .setBody(toSend.hasArray()
```

Comment:
this can be:

```
if (toSend.hasArray) {
    UnsafeByteOperations.unsafeWrap(toSend.array(), toSend.arrayOffset(), toSend.readableBytes());
} else if (toSend.size() == 1) { // only one bytebuffer
   UnsafeByteOperations.unsafeWrap(toSend.getBuffer(0).nioBuffer());
} else {
   UnsafeByteOperations.unsafeWrap(toSend.toArray());
}
```

Reply:
Ok.